demos/guide/quic-server-non-block.c

```diff
@@ -0,0 +1,495 @@
+            break;
+
+        default:
+            printf("Unknown stream failure\n");
+            break;
+        }
+        /*
+         * If the failure is due to a verification error we can get more
+         * information about it from SSL_get_verify_result().
+         */
+        if (SSL_get_verify_result(ssl) != X509_V_OK)
+            printf("Verify error: %s\n",
+                   X509_verify_cert_error_string(SSL_get_verify_result(ssl)));
+        return -1;
+
+    default:
+        return -1;
+    }
+}
+
+/*
+ * Main loop for server to accept QUIC connections.
+ * Echo every request back to the client.
+ */
+static int run_quic_server(SSL_CTX *ctx, int fd)
+{
+    int ok = -1;
+    int ret, eof;
+    SSL *listener, *conn = NULL;
+    unsigned char buf[8192];
+    size_t nread, nwritten, total_read, total_written;
+
+    /* Create a new QUIC listener */
+    if ((listener = SSL_new_listener(ctx, 0)) == NULL)
+        goto err;
+
+    /* Provide the listener with our UDP socket. */
+    if (!SSL_set_fd(listener, fd))
+        goto err;
+
+    /*
+     * Set the listener mode to non-blocking, which is inherited by
+     * child objects.
+     */
+    if (!SSL_set_blocking_mode(listener, 0))
+        goto err;
+
+    /* Begin listening. */
+    if (!SSL_listen(listener))
+        goto err;
+
+    /*
+     * Begin an infinite loop of listening for connections. We will only
+     * exit this loop if we encounter an error.
+     */
+    for (;;) {
+        eof = 0;
+        total_read = 0;
+        total_written = 0;
+
+        /* Pristine error stack for each new connection */
+        ERR_clear_error();
+
+        /* Block while waiting for a client connection */
+        printf("Waiting for connection\n");
+        while ((conn = SSL_accept_connection(listener, 0)) == NULL)
+            wait_for_activity(listener);
+        printf("Accepted new connection\n");
+
+        /* Read from client until the client sends a end of stream packet */
+        while (!eof) {
+            ret = SSL_read_ex(conn, buf + total_read, sizeof(buf) - total_read,
+                              &nread);
+            total_read += nread;
+            if (total_read >= 8192) {
+                fprintf(stderr, "Could not fit all data into buffer\n");
+                goto err;
+            }
+
+            switch (handle_io_failure(conn, ret)) {
+            case 1:
+                continue; /* Retry */
+            case 0:
+                /* Reached end of stream */
+                if (!SSL_has_pending(conn))
+                    eof = 1;
+                break;
+            default:
+                fprintf(stderr, "Failed reading remaining data\n");
+                goto err;
+            }
+        }
+
+        /* Echo client input */
+        while (!SSL_write_ex2(conn, buf + total_written,
+                              total_read - total_written,
+                              SSL_WRITE_FLAG_CONCLUDE, &nwritten)) {
```

Comment:
Sorry - one last comment!!

`total_written` will always be 0 here so it seems to me this could be simplified as:

```suggestion
        while (!SSL_write_ex2(conn, buf,
                              total_read,
                              SSL_WRITE_FLAG_CONCLUDE, &total_written)) {
```
Which means we can totally do away with `nwritten` altogether